### ISSUE TITLE
xgboost4j spark supports incremental learning

### ISSUE THREAD

**Issue by zyq11223** [dmlc/xgboost#4188]
how to incremental training xgboost4j  spark model  with an existed xgboost classification model ？
Does the XGBModel.booster.update(DMatrix, Iteration ) mean that ?

**Comment by CodingCat** [comment]
I think only incremental training with the dataset (and with the same partitioning schema in spark case) is theoretically correct....right? @hcho3

**Comment by hcho3** [comment]
@CodingCat I don't think we provide a theoretically robust method for incremental training. Does anyone manage to do that?

**Comment by CodingCat** [comment]
No, though we have a checkpoint feature in spark, I am not sure if changing the partition of dataset will lead to error in theory....

**Comment by hcho3** [comment]
@CodingCat I see, yes, with checkpointing, you'd need to have the same training data to get robust result. I'm not sure whether change in partition would be a problem itself, as long as the whole dataset is the same. On the other hand, if you change the training data itself (this is what people mean by streaming/incremental training), then you throw all guarantees out of the window.

**Comment by CodingCat** [comment]
my worry is that change of partition would potentially change the synced sketch and therefore histogram buckets, 

Then your booster are going through two different sets of potential splits...

But I cannot really prove it in math

**Comment by hcho3** [comment]
@CodingCat We should add checkpointing for the histogram sketch stage then?

**Comment by CodingCat** [comment]
@hcho3 I think we should extend the checkpoint file format in Spark to record partitioning information (e.g. number of partitions, partitioner, etc.)

but I think this issue should be non-severe  with the increasing of dataset size.....so not urgent to bring that fix

anyway let's close this issue, conclusion is, `NO, gradient boosting does not allow "incremental" training`

**Comment by zyq11223** [comment]
@CodingCat
 Could you mind  telling me that why xgboost4j python supports incremental learning with an existed ? And xgboost4j python has the interface which can train incrementally .Thanks !

And another questions. I used xgbModel.booster.update API of xgboost spark , it just updates one tree ,  and  with trying to set related params (
 xgbModel.booster.setParam("process_type", "update")
    xgbModel.booster.setParam("updater", "refresh")
    xgbModel.booster.setParam("refresh_leaf", "1")
).
It just also can not update all the trees of the existed model. Are there some api or interface that could solve it ?
Appreciate!!
@hcho3

**Comment by CodingCat** [comment]
if I understand your words correctly, you say python supports incremental training, how?

**Comment by zyq11223** [comment]
> if I understand your words correctly, you say python supports incremental training, how?

here:
one_shot_model_itr = xgb.train({
    'updater':'refresh',
    'process_type': 'update',
    'refresh_leaf': True,
    'silent': False,
    'max_depth':3,
    'objective':'reg:linear'
},num_boost_round=10, dtrain=xgb.DMatrix(x_te, y_te), xgb_model=one_shot_model)

or 
the links below:
https://stackoverflow.com/questions/38079853/how-can-i-implement-incremental-training-for-xgboost

**Comment by zyq11223** [comment]
> > if I understand your words correctly, you say python supports incremental training, how?
> 
> here:
> one_shot_model_itr = xgb.train({
> 'updater':'refresh',
> 'process_type': 'update',
> 'refresh_leaf': True,
> 'silent': False,
> 'max_depth':3,
> 'objective':'reg:linear'
> },num_boost_round=10, dtrain=xgb.DMatrix(x_te, y_te), xgb_model=one_shot_model)
> 
> and
> the links below:
> 
> https://stackoverflow.com/questions/38079853/how-can-i-implement-incremental-training-for-xgboost

train on new dataset with an existed model

**Comment by CodingCat** [comment]
so, we have said that gradient boosting does not allow it in math

check the discussion here, https://github.com/dmlc/xgboost/issues/3055

**Comment by zyq11223** [comment]
> so, we have said that gradient boosting does not allow it in math
> 
> check the discussion here, #3055

Ok, so xgboost4j spark does not provide api for incremental training , and i am very curiosity about why python supports incremental training like the way above.
Thank you a lot !

**Comment by CodingCat** [comment]
................................................


please read #3055, no language binding really supports incremental training

the thing you post here is a "undefined behavior" in xgboost, sometimes it just works, sometimes not, because there is not math back it....

**Comment by zyq11223** [comment]
> ................................................
> 
> please read #3055, no language binding really supports incremental training
> 
> the thing you post here is a "undefined behavior" in xgboost, sometimes it just works, sometimes not, because there is no math back it....

Very Appreciate for your reply.

**Comment by hcho3** [comment]
@CodingCat Indeed, to perform "incremental training" in mathematically robust way, we'd need to implement a new algorithm. Actually, this new algorithm would be really useful in other scenarios, e.g. "dynamic training," where training may proceed with a varying number of workers over time (Spot Fleet).
